File: leetcode/0007-reverse-integer/main.R
reverse <- function(x) {
  if (x >= 0) {
    tmp <- 0
    # R 的求余和求除操作与 Go 不同
    while (x != 0) {
      tmp <- tmp * 10 + x %% 10
      x <- x %/% 10
    }
  } else {
    tmp <- -reverse(-x)
  }

  if (tmp < -2^31 || tmp > 2^31 - 1) tmp <- 0
  tmp
}

cat(reverse(123), "\n")
cat(reverse(-123), "\n")
cat(reverse(120), "\n")
cat(reverse(0), "\n")

# r$> -123 %% 10
# [1] 7

File: leetcode/0007-reverse-integer/main.go
package main

import "fmt"

func reverse(x int) int {
	tmp := 0
	for x != 0 {
		// 使用一个临时变量逐步存储逆转的数值
		tmp = tmp*10 + x%10
		x = x / 10
	}

	if tmp > 1<<31-1 || tmp < -(1<<31) {
		return 0
	}
	return tmp
}

func main() {
	fmt.Println(reverse(123))
	fmt.Println(reverse(-123))
	fmt.Println(reverse(0))
}
